### DISCUSSION TITLE
Bug in UART Mux?

### DISCUSSION THREAD

**Discussion by AeroSynth** [micropython/micropython#13102] [ESP32]
Possible Bug
I’ve found a relatively nasty issue with the UART Mux on the ESP32-S3.  I am testing this on a ESP32-S3-DevKitC-1 P2N8 and will eventually use the ESP32-S3-Mini.
I can assign UART0/1/2 to alternate pins.  That works fine.
But the original IO pins are NOT disconnected.  For example, assigning UART2 to IO39 (Tx) and IO38 (Rx) works fine.  But the UART data also appears at the default IO17 and IO16.  And this is particularly troubling as I plan to connect IO16 and IO15 to an external 32KHz crystal as per the data sheet.

It’s the same with UART0 and UART1.  No matter which IO pin I use, the default IO remains.
Is this a bug, or is there something I'm missing?

There also seems to be a problem with uart.deinit() with Thonny.  It throws an exception every time.  Annoying, but I’m not using deinit at the moment.

**Comment by AeroSynth** [comment]
It turns out that @xyzzy42 had a patch to uart_machine.c that may fix the problem.  I'll test it out.
https://github.com/xyzzy42/micropython/commit/4019e081dbc01a704c55779c92590a72c9581d9a

**Comment by trentp-igor** [comment]
This is likely a bug with every ESP32 peripheral supported by Micropython.

The standard for ESP-IDF drivers is that `xxxx_set_pin()` does **not** disconnect the previously used pins from the function.  This is obviously not a good design, but Espressif knows about it.  They must think it's ok, or don't want to admit it's a bad design, or are afraid of breaking code that relies on this behavior (I doubt any such code exists), or something?  Boggles the mind, but that's how it is.

It's not as huge a problem as it might seem, because using the pins for another peripheral will reconfigure them.  And if the pins don't get re-used, then it doesn't break anything that there is an extra UART output or whatever on them.  There are probably a ton of ESP32 based devices out there spewing EMI from unconnected pins that still have a signal on them.

It's a problem here because there's no xtal peripheral that will be initialized and configure the pins to the correct mode.
